### PR TITLE
docker support

### PR DESCRIPTION
**Pull request by shariq1501**

### CODE REVIEW

File: src/frontend/README.md [hunk 1]
```diff
@@ -62,10 +62,27 @@ you should be able to access it on 0.0.0.0:8080
 ## REST API URL Configuration:
 
 ### Dev Mode:
-Set api base URL in `environments/environment.ts` file  
+Set default API base URL in `environments/environment.ts`, you can always change it and restart the from settings.
```
Comment (praveen-garg) [inline src/frontend/README.md]: restart the app, from settings.

File: src/frontend/README.md [hunk 2]
```diff
@@ -62,10 +62,27 @@ you should be able to access it on 0.0.0.0:8080
 ## REST API URL Configuration:
 
 ### Dev Mode:
-Set api base URL in `environments/environment.ts` file  
+Set default API base URL in `environments/environment.ts`, you can always change it and restart the from settings. 
 
 ### Production Mode:
-Set api base URL in `environments/environment.prod.ts` file  
+Set API base URL in `environments/environment.prod.ts`, you can always change it and restart the from settings.
```
Comment (praveen-garg) [inline src/frontend/README.md]: restart the app, from settings.

File: src/frontend/README.md [hunk 3]
```diff
@@ -62,10 +62,27 @@ you should be able to access it on 0.0.0.0:8080
...
+Set API base URL in `environments/environment.prod.ts`, you can always change it and restart the from settings. 
+
+### Running using Docker 
+1) Install Docker CE for Mac or Windows (http://docker.com)
+
+2) git clone https://github./praveen-garg/foglamp/FogLAMP
+
+3) Move to `src/frontend` directory and run `yarn install`.
+
+4) Run `yarn build`
```
Comment (praveen-garg) [inline src/frontend/README.md]: I assume all above are common!

We need to add `Prerequisite ` section too; node, npm, yarn, cli etc.

File: src/frontend/README.md [hunk 4]
```diff
@@ -80,10 +97,10 @@ Set api base URL in `environments/environment.prod.ts` file
...
  Library      |   Version     | Latest Stable (? Y/n) | License
 ------------- | ------------- | --------------------  | ------------
-canvas-gauges |  2.1.4        |        Yes            | MIT 
-chart.js      |  2.7.0        |        Yes            | MIT 
-core-js       |  2.5.1        |        Yes            | MIT 
-loadash       |  4.17.4       |        Yes            | MIT
-ngx-mask      |  1.0.3        |        Yes            | MIT 
-ng-sidebar    |  6.0.2        |        Yes            | MIT 
-moment        |  2.19.2       |        Yes            | MIT
+canvas-gauges |  2.1.4        |        2.1.4          | MIT
```
Comment (praveen-garg) [inline src/frontend/README.md]: please add (TODO: remove) some-where in this line

File: src/frontend/docker/nginx.dockerfile [hunk 5]
```diff
@@ -0,0 +1,10 @@
+FROM nginx:alpine
+
+LABEL author="Mohd. Shariq" 
+
+# Copy custom nginx config
+COPY ./nginx.conf /etc/nginx/nginx.conf
+
+EXPOSE 8080
+
+ENTRYPOINT ["nginx", "-g", "daemon off;"]
```
Comment (praveen-garg) [inline src/frontend/docker/nginx.dockerfile]: daemon off? can we add a comment in file? and what is the other option? when to use what?

File: src/frontend/nginx.conf [hunk 6]
```diff
@@ -1,32 +1,25 @@
-events {
-  worker_connections  256;
-}
+worker_processes 4;
```
Comment (praveen-garg) [inline src/frontend/nginx.conf]: how it will help?

File: src/frontend/nginx.conf [hunk 7]
```diff
@@ -1,32 +1,25 @@
...
+
+events { worker_connections 1024; }
```
Comment (praveen-garg) [inline src/frontend/nginx.conf]: how it will help here?

File: src/frontend/nginx.conf [hunk 8]
```diff
@@ -1,32 +1,25 @@
...
+      listen       80;
+      server_name  localhost;
 
-        # serve static content
-        location /static/ {
-            # path for static files
-            root .;
-        }
+      location / {
+          root   /usr/share/nginx/html;
```
Comment (praveen-garg) [inline src/frontend/nginx.conf]: is this part of server or location?

File: src/frontend/nginx.conf [hunk 9]
```diff
@@ -1,32 +1,25 @@
...
+          index  index.html;
```
Comment (praveen-garg) [inline src/frontend/nginx.conf]: indent properly as per sibling node

File: src/frontend/nginx.conf [hunk 10]
```diff
@@ -1,32 +1,25 @@
...
+          try_files $uri$args $uri$args/ $uri $uri/ /index.html =404;
```
Comment (praveen-garg) [inline src/frontend/nginx.conf]: how it will help?

File: src/frontend/nginx.conf [hunk 11]
```diff
@@ -1,32 +1,25 @@
...
 http {
     include   /etc/nginx/mime.types; # fix it as per your env
+       
     server {
-        root dist/; # necessary to avoid nginx injecting root into try_files directives
-
-        listen 8080;
-        server_name localhost;
-        server_name 127.0.0.1;
-        sendfile off;
```
Comment (praveen-garg) [inline src/frontend/nginx.conf]: why these 2 removed? `server_name 127.0.0.1;`  `sendfile off;`

File: src/frontend/nginx.conf (hunk 8 above)
Comment (praveen-garg) [reply src/frontend/nginx.conf]: and will this allow to run withOUT docker? (as we used to run of any OS with this conf file on mac?

**Review by praveen-garg — CHANGES_REQUESTED** [review]

File: src/frontend/docker/nginx.dockerfile (hunk 5 above)
Comment (shariq1501) [reply src/frontend/docker/nginx.dockerfile]: nginx can be configured not to self-daemonize by passing `daemon off`.

*nginx doc says* 
```
Syntax:  daemon on | off
Default:  daemon on;
```
Determines whether nginx should become a daemon. Mainly used during development.

But still if you want to run in deamon mode run ```docker-compose up -d```

File: src/frontend/nginx.conf (hunk 7 above)
Comment (shariq1501) [reply src/frontend/nginx.conf]: Sets the maximum number of simultaneous connections that can be opened by a worker process.
http://nginx.org/en/docs/ngx_core_module.html#worker_connections

File: src/frontend/nginx.conf [hunk 12]
```diff
@@ -1,32 +1,25 @@
-events {
-  worker_connections  256;
-}
+worker_processes 4;
+
+events { worker_connections 1024; }
 
 http {
     include   /etc/nginx/mime.types; # fix it as per your env
```
Comment (praveen-garg) [inline src/frontend/nginx.conf]: for docker, we need this? we should not say `# fix it as per your env` for docker

Comment (praveen-garg) [reply src/frontend/nginx.conf]: let's have 2 conf -- `nginx.conf` and `nginx-docker.conf` should have very minimal difference to make it work

**Review by praveen-garg — APPROVED** [review]

**Pull request merged**